app/src/main/java/org/schabi/newpipe/util/NewPipeTextViewHelper.java

```diff
@@ -0,0 +1,81 @@
+package org.schabi.newpipe.util;
+
+import android.content.Context;
+import android.text.Selection;
+import android.text.Spannable;
+import android.widget.TextView;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import org.schabi.newpipe.util.external_communication.ShareUtils;
+import org.schabi.newpipe.views.NewPipeEditText;
+import org.schabi.newpipe.views.NewPipeTextView;
+
+public final class NewPipeTextViewHelper {
+    private NewPipeTextViewHelper() {
+    }
+
+    /**
+     * Share the selected text of {@link NewPipeTextView NewPipeTextViews} and
+     * {@link NewPipeEditText NewPipeEditTexts} with
+     * {@link ShareUtils#shareText(Context, String, String)}.
+     *
+     * <p>
+     * This allows EMUI users to get the Android share sheet instead of the EMUI share sheet when
+     * using the {@code Share} command of the popup menu which appears when selecting text.
+     * </p>
+     *
+     * @param textView the {@link TextView} on which sharing the selected text. It should be a
+     *                 {@link NewPipeTextView} or a {@link NewPipeEditText} (even if
+     *                 {@link TextView standard TextViews} are supported).
+     */
+    public static void shareSelectedTextWithShareUtils(@NonNull final TextView textView) {
+        if (!(textView instanceof NewPipeEditText)) {
+            final CharSequence textViewText;
+            if (textView instanceof NewPipeTextView) {
+                final NewPipeTextView newPipeTextView = (NewPipeTextView) textView;
+                textViewText = newPipeTextView.getText();
+            } else {
+                textViewText = textView.getText();
+            }
+
+            final CharSequence selectedText = getSelectedText(textView, textViewText);
+            shareSelectedTextIfNotNullAndNotEmpty(textView, selectedText);
+
+            final Spannable spannable = (textViewText instanceof Spannable)
+                    ? (Spannable) textViewText : null;
+            if (spannable != null) {
+                Selection.setSelection(spannable, textView.getSelectionEnd());
+            }
+        } else {
+            final NewPipeEditText editText = (NewPipeEditText) textView;
+            final Spannable text = editText.getText();
+
+            final CharSequence selectedText = getSelectedText(textView, text);
+            shareSelectedTextIfNotNullAndNotEmpty(textView, selectedText);
+            Selection.setSelection(text, editText.getSelectionEnd());
+        }
+    }
```

Comment:
Simplifies code:
```suggestion
    /**
     * Shares the selected text of a {@link TextView} with
     * {@link ShareUtils#shareText(Context, String, String)}.
     *
     * <p>
     * This allows EMUI users to get the Android share sheet instead of the EMUI share sheet when
     * using the {@code Share} command of the popup menu which appears when selecting text.
     * </p>
     *
     * @param textView the {@link TextView} on which sharing the selected text. It should be a
     *                 {@link NewPipeTextView} or a {@link NewPipeEditText} (even if
     *                 {@link TextView standard TextViews} are supported).
     */
    public static void shareSelectedTextWithShareUtils(@NonNull final TextView textView) {
        final CharSequence textViewText = textView.getText();

        shareSelectedTextIfNotNullAndNotEmpty(
                textView,
                getSelectedText(textView, textViewText)
        );

        if (textViewText instanceof Spannable) {
            Selection.setSelection((Spannable) textViewText, textView.getSelectionEnd());
        }
    }
```

Reply:
The only problem of your suggestion is that I would not recommend using the getText method of the `TextView` class when the `TextView` is a `NewPipeTextView` or a `NewPipeEditText`.

Why? Because these classes extends from AndroidX classes, which adds some code to their implementation of the `getText` method:

- for `AppCompatTextView`, see here: https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-main/appcompat/appcompat/src/main/java/androidx/appcompat/widget/AppCompatTextView.java#550
- for `AppCompatEditText`, see here: https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-main/appcompat/appcompat/src/main/java/androidx/appcompat/widget/AppCompatEditText.java#120

Not using these methods may result in crashes or bugs in the app, depending of devices/Android versions or of the context when the method is called.

Reply:
> The only problem of your suggestion is that I would not recommend using the getText method of the TextView class when the TextView is a NewPipeTextView or a NewPipeEditText.

![grafik](https://user-images.githubusercontent.com/40789489/135768699-e8a72be2-711c-49a2-a6d7-90b43c6639f5.png)
<sub>First thought: Am I getting something wrong?</sub>
But that's not how inheritance works (in Java):

Java calls the method ``getText`` on the passed object and it uses the type (at runtime) of the passed object.
So if you e.g. supply a ``NewPipeTextView`` as ``textView`` it will still call the ``getText`` of ``NewPipeTextView``.

https://user-images.githubusercontent.com/40789489/135768780-4aeb7b9a-ccc4-4c4f-b1eb-827900ddf182.mp4